### ISSUE TITLE
Add Bounding Box to Differencing Video Filter

### ISSUE THREAD

**Issue by E3t1jump** [Engineering-3/RCM-BFin#6]
We want to see if we can get the bounding box tool to work within the Differencing video filter. The foundation goal of the bounding box use here is to demonstrate the video filter can act like a sensor by A) Detecting an event, B) Demonstrating what is detected, and C) triggering a response (in our first case we can save video frames to file, but we may need to be able to trigger other responses at some point).

**Comment by EmbeddedMan** [comment]
When playing with the existing differencing function (g0), I'm struggling a bit to figure out how to put a bounding box around all of the differences. Say for example I put my hand in front of the camera, then turn on differencing. The BFin takes a single frame at that point in time, and then from that point on only shows frames of the difference between the current frame and that first captured frame. So if I move my hand, everywhere I move my hand (and the spot where my hand was originally) is shown in the frames.

Maybe we can use the threshold value, and draw bounding boxes around each area that has pixels brighter than the threshold. Does that seem reasonable?

The difficult part will be breaking up the thresholded image into all of the various sub-rectangles that indicate each part of the image that's different from the first frame. There are some algorithms out there which will allow you to group parts of an image together (in this case by brightness) and draw bounding boxes around each group. I'll read up on those.

**Comment by E3t1jump** [comment]
Yes, I was concerned about how to isolate motion in such a large format.

To expand a bit: My first direction with motion detection was to use the Blob Detection mode. Blob detection already identifies with a bounding box, and if the blob can be detected, the bounding box will follow the blob. The key with blob detection is setting a color range to detect. The problem we were having in AUS was that the target blobs to detect were black, as was a multitude of other items within the viewing field. The target blobs are also very small (4mm and 1mm) so the  bounding boxes would grab large areas with no isolation on the target blobs.

When we switched to the Differencing filter we could see both the 4mm and 1mm blobs moving with ease. The Differencing filter was only motion sensitive and not related to color.

The rules state we must place a bounding box around the moving object and the box must track with the object to show there is some "sensing" action taking place. Seeing the motion by the operator is not enough. The camera needs to detect the motion (aka machine vision, and demonstration that there is vision processing taking place that can be used to trigger a response without operator intervention). Maybe??? if we can trigger some action when motion is detected through the Differencing filter, we can argue against the bounding box requirement in favor of alternate methods of proving our system is meeting the real goal vs. a specific identification process.

Alternately, if we need to supply the bounding box, maybe we only box in small items and not something the size of a hand?? To note, when the robot is driving around every full image is illuminated because the camera is moving. I only see this working if we activate the filter after we've come to a full stop.

Thoughts?

**Comment by EmbeddedMan** [comment]
OK, I'll give it a try. I'll create a 1bpp image based on the threshold (just like for QR code detection) based on the difference frames, and then draw bounding boxes around each 'blob' that I see. I think I can actually rework the blob detection code that's already there to do both things. This may be a dead end, but we'll give it a try and see how it works for you.
